exporter/splunkhecexporter/logdata_to_splunk.go

```diff
@@ -15,29 +15,162 @@
+
+		// chunk buffers events up to the max content length.
+		chunk := &eventsChunk{buf: new(bytes.Buffer)}
+
+		rl := ld.ResourceLogs()
+		for i := 0; i < rl.Len(); i++ {
+			ill := rl.At(i).InstrumentationLibraryLogs()
+			for j := 0; j < ill.Len(); j++ {
+				l := ill.At(j).Logs()
+				for k := 0; k < l.Len(); k++ {
+					if chunk.index == nil {
+						chunk.index = &logIndex{origIdx: i, instIdx: j, logsIdx: k}
+					}
+
+					if err := encoder.Encode(mapLogRecordToSplunkEvent(l.At(k), config, logger)); err != nil {
+						select {
+						case <-ctx.Done():
+						case chunkCh <- &eventsChunk{buf: nil, index: chunk.index, err: consumererror.Permanent(err)}:
+						}
+						return
+					}
+					event.WriteString("\r\n\r\n")
+
+				addToChunk:
+					// The size of an event must be less than or equal to max content length.
+					if config.MaxContentLengthLogs > 0 && event.Len() > config.MaxContentLengthLogs {
+						select {
+						case <-ctx.Done():
+						case chunkCh <- &eventsChunk{
+							buf:   nil,
+							index: chunk.index,
+							err:   consumererror.Permanent(fmt.Errorf("log event bytes exceed max content length configured (log: %d, max: %d)", event.Len(), config.MaxContentLengthLogs))}:
+						}
+						return
+					}
+
+					// Moving the event to chunk.buf if length will be <= max content length.
+					// Max content length <= 0 is interpreted as unbound.
+					if chunk.buf.Len()+event.Len() <= config.MaxContentLengthLogs || config.MaxContentLengthLogs <= 0 {
+						// WriteTo() empties and resets buffer event.
+						if _, err := event.WriteTo(chunk.buf); err != nil {
+							select {
+							case <-ctx.Done():
+							case chunkCh <- &eventsChunk{buf: nil, index: chunk.index, err: consumererror.Permanent(err)}:
+							}
+							return
+						}
+						continue
+					}
+
+					select {
+					case <-ctx.Done():
+						return
+					case chunkCh <- chunk:
+					}
+
+					// Creating a new events buffer.
+					chunk = &eventsChunk{buf: new(bytes.Buffer)}
+
+					// Adding remaining event to the new chunk
+					if event.Len() != 0 {
+						chunk.index = &logIndex{origIdx: i, instIdx: j, logsIdx: k}
+						goto addToChunk
+					}
+				}
+			}
+		}
+
+		select {
+		case <-ctx.Done():
+			return
+		case chunkCh <- chunk:
+		}
+	}()
+
+	return chunkCh
+}
+
+func (ld *logDataWrapper) numLogs(from *logIndex) int {
+	count, orig := 0, *ld.InternalRep().Orig
+
+	// Validating logIndex. Invalid index will cause out of range panic.
+	_ = orig[from.origIdx].InstrumentationLibraryLogs[from.instIdx].Logs[from.logsIdx]
+
+	for i := from.origIdx; i < len(orig); i++ {
+		for j, library := range orig[i].InstrumentationLibraryLogs {
+			switch {
+			case i == from.origIdx && j < from.instIdx:
+				continue
+			default:
+				count += len(library.Logs)
 			}
 		}
 	}
 
-	return splunkEvents
+	return count - from.logsIdx
+}
+
+func (ld *logDataWrapper) subLogs(from *logIndex) *pdata.Logs {
+	clone := ld.Clone().InternalRep()
```

Comment:
We ideally shouldn't be digging into the internal protobuf. I'll see if this is doable with current pdata interfaces or if something new is needed there.

Reply:
I haven't had a chance to look into this. Can you see what @bogdandrutu thinks as I think he might have already hidden InternalRep in contrib.